File: cpp/cpp_10-99/56.cc
// C++
// leetcode 56
// https://leetcode-cn.com/problems/merge-intervals/
// 先排序，然后合并


class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());

        vector< vector<int>> res;
        for(int i = 0; i< intervals.size(); ){
            int j = i+1;
            int max = intervals[i][1];
            while(j < intervals.size() && intervals[j][0] <= max){
                if(max < intervals[j][1])
                	max = intervals[j][1];
                j++;
            }
            res.push_back({intervals[i][0], max});
            i = j;
        }
       
        return res;
    }
};

File: Go/Go_10-99/56.go
// Go
// leetcode 56
// https://leetcode.cn/problems/merge-intervals/

func merge(intervals [][]int) (res [][]int) {
	// 排序
	sort.Slice(intervals, func(i, j int) bool {
		a, b := intervals[i], intervals[j]
		if a[0] != b[0] {
			return a[0] < b[0]
		}
		return a[1] < b[1]
	})

	// fmt.Println(intervals)
	preA, preB := intervals[0][0], intervals[0][1]
	for i := 1; i < len(intervals); i++ {
		a, b := intervals[i][0], intervals[i][1]
		if a <= preB {
			preB = max(b, preB)
			continue
		}

		res = append(res, []int{preA, preB})
		preA, preB = a, b
	}
	res = append(res, []int{preA, preB})

	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}